File: src/validation/stage3.ts
import type { ValidationIssue } from "../config.js";

import { checkField, readFileOrIssues } from "./common.js";

const REQUIRED_FIELDS = [
  "Type",
  "Location",
  "Attacker-controlled data",
  "Analysis hints",
];
const VALID_TYPES = new Set(["p", "h", "s", "parser", "handler", "session"]);

function splitEntryPoints(content: string): Array<{ epId: string; block: string }> {
  const splits = content.split(/###\s+EP-(\d+)\s*:/);
  const results: Array<{ epId: string; block: string }> = [];

  for (let index = 1; index < splits.length - 1; index += 2) {
    const number = splits[index];
    const block = splits[index + 1];
    if (!number || block === undefined) {
      continue;
    }
    results.push({ epId: `EP-${number}`, block });
  }

  return results;
}

export function validateStage3File(filePath: string): ValidationIssue[] {
  const { content, issues } = readFileOrIssues(filePath);
  if (issues.length > 0) {
    return issues;
  }

  if (!content.trim()) {
    return [
      {
        description: "Output file is empty.",
        expected: "At least one entry point block (### EP-{N}:) with required fields.",
        fix: "Write the identified entry points to this file.",
      },
    ];
  }

  const entryPoints = splitEntryPoints(content);
  if (entryPoints.length === 0) {
    return [
      {
        description: "No entry point blocks found.",
        expected: 'At least one "### EP-{N}:" block (e.g., "### EP-1:").',
        fix: 'Add entry point blocks using the format "### EP-1:" followed by the required fields.',
      },
    ];
  }

  const validationIssues: ValidationIssue[] = [];
  for (const { epId, block } of entryPoints) {
    for (const field of REQUIRED_FIELDS) {
      const value = checkField(block, field);
      if (value === null) {
        validationIssues.push({
          description: `${epId}: Missing required field "**${field}**".`,
          expected: `Each entry point must have a "- **${field}**: ..." line.`,
          fix: `Add "- **${field}**: <value>" to the ${epId} block.`,
        });
      } else if (!value || ["none", "n/a", "..."].includes(value.toLowerCase())) {
        if (field === "Type" || field === "Location") {
          validationIssues.push({
            description: `${epId}: Field "**${field}**" has placeholder or empty value: "${value}".`,
            expected: `A concrete value for ${field}.`,
            fix: `Fill in the actual ${field} for ${epId}.`,
          });
        }
      }
    }

    const typeValue = checkField(block, "Type");
    if (typeValue) {
      const typeToken = typeValue.split(/\s+/, 1)[0]?.replace(/[()]/g, "").toLowerCase() ?? "";
      if (!VALID_TYPES.has(typeToken)) {
        validationIssues.push({
          description: `${epId}: Invalid Type value "${typeValue}".`,
          expected: 'Type must be one of: P (Parser), H (Handler), S (Session).',
          fix: `Change the Type of ${epId} to "P (Parser)", "H (Handler)", or "S (Session)".`,
        });
      }
    }
  }

  return validationIssues;
}

File: code_auditor/validation/stage3.py
from __future__ import annotations

import json

from ..config import ValidationIssue
from .common import read_file_or_issues

_PLACEHOLDERS = {"none", "n/a", "...", "tbd", ""}


def _is_blank(value: object) -> bool:
    if isinstance(value, str):
        return value.lower().strip() in _PLACEHOLDERS
    if isinstance(value, list):
        return len(value) == 0
    return not value


def validate_stage3_file(file_path: str) -> list[ValidationIssue]:
    content, issues = read_file_or_issues(file_path)
    if issues:
        return issues

    if not content.strip():
        return [ValidationIssue(
            description="Analysis unit file is empty.",
            expected="A JSON object with description, files, and focus fields.",
            fix="Write the analysis unit definition as JSON to this file.",
        )]

    try:
        data = json.loads(content)
    except json.JSONDecodeError as e:
        return [ValidationIssue(
            description=f"Invalid JSON: {e}",
            expected="Valid JSON.",
            fix="Fix the JSON syntax error.",
        )]

    validation_issues: list[ValidationIssue] = []

    if _is_blank(data.get("description")):
        validation_issues.append(ValidationIssue(
            description='Missing or blank "description".',
            expected="A short description of what this unit covers.",
            fix='Add a "description" field.',
        ))
    if _is_blank(data.get("files")):
        validation_issues.append(ValidationIssue(
            description='Missing or empty "files".',
            expected="A non-empty array of source file paths.",
            fix='Add a "files" array with at least one path.',
        ))
    if _is_blank(data.get("focus")):
        validation_issues.append(ValidationIssue(
            description='Missing or blank "focus".',
            expected="Concrete analysis guidance.",
            fix='Add a "focus" field with actionable analysis guidance.',
        ))

    return validation_issues
